botorch/acquisition/hentropy_search.py

```diff
@@ -0,0 +1,282 @@
+                resample=False,
+                collapse_batch_dims=True,
+            )
+        elif n_fantasy_at_action_pts is not None:
+            if action_sampler.sample_shape != torch.Size([n_fantasy_at_action_pts]):
+                raise ValueError(
+                    "The sampler shape must match n_fantasy_at_action_pts="
+                    f"{n_fantasy_at_action_pts}."
+                )
+        else:
+            n_fantasy_at_action_pts = action_sampler.sample_shape[0]
+
+        self.design_sampler = design_sampler
+        self.action_sampler = action_sampler
+        self.n_fantasy_at_design_pts = n_fantasy_at_design_pts
+        self.loss_function_hyperparameters = loss_function_hyperparameters
+        self.loss_function = loss_function_class(
+            **self.loss_function_hyperparameters,
+        )
+
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor, A: Tensor) -> Tensor:
+        r"""Evaluate qHEntropySearch objective (q-HES) on the candidate set `X`.
+
+        Args:
+            X: Design tensor of shape `(batch) x q x num_dim_design`.
+            A: Action tensor of shape `(batch) x n_fantasy_at_design_pts
+                x num_actions x num_dim_action`.
+
+        Returns:
+            A Tensor of shape `(batch)`.
+        """
+
+        # construct the fantasy model of shape `n_fantasy_at_design_pts x b`
+        fantasy_model = self.model.fantasize(
+            X=X, sampler=self.design_sampler, observation_noise=True
+        )
+
+        # Permute shape of A to work with self.model.posterior correctly
+        A = A.permute(1, 0, 2, 3)
+
+        fantasized_outcome = self.action_sampler(fantasy_model.posterior(A))
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions x 1
+
+        fantasized_outcome = fantasized_outcome.squeeze(dim=-1)
+        # >>> n_fantasy_at_action_pts x n_fantasy_at_design_pts
+        # ... x batch_size x num_actions
+
+        values = self.loss_function(A=A, Y=fantasized_outcome)
+        # >>> n_fantasy_at_design_pts x batch_size
+
+        # return average over the fantasy samples
+        return values.mean(dim=0)
+
+    def get_augmented_q_batch_size(self, q: int) -> int:
+        r"""Get augmented q batch size for optimization.
+
+        Args:
+            q: The number of candidates to consider jointly.
+
+        Returns:
+            The augmented size for optimization.
+        """
+
+        return q + self.cfg.n_dim_action
+
+    def extract_candidates(self, batch_as_full: Tensor) -> Tensor:
+        r"""We only return X as the set of candidates post-optimization.
+
+        Args:
+            batch_as_full: A `b x (q + num_fantasies) x d`-dim Tensor with `b`
+                t-batches of `q + num_fantasies` design points each.
+
+        Returns:
+            A `b x q x d`-dim Tensor with `b` t-batches of `q` design points each.
+        """
+
+        assert len(batch_as_full.shape) == 2
```

Comment:
Is the docstring wrong? The docstring makes it sound like this would be 3d.

Also, we don't usually use asserts -- something like a `ValueError` would be better. If you were to add a ValueError here, you'd need to add a unit test for it. You could also just remove this line.

Reply:
The output shape should be 3D. I adopt the docstring from the MST file: https://github.com/pytorch/botorch/blob/main/botorch/acquisition/multi_step_lookahead.py#L283